charms/knative-operator/src/charm.py

```diff
@@ -142,6 +186,29 @@ def _on_knative_operator_pebble_ready(self, event):
         self.unit.status = MaintenanceStatus("Configuring Pebble layer")
         self._update_layer(event)
 
+    def _on_otel_collector_relation_changed(self, event):
+        """Event handler for on['otel-collector'].relation_changed."""
+        # Apply all changes only if the otel collector has not been deployed
+        if not self._otel_exporter_ip:
+            self.unit.status = MaintenanceStatus("Applying otel collector")
+            self._template_files_ext = [
+                f"{self._src_dir}/manifests/observability/collector.yaml.j2"
+            ]
+            # Reset the resource handler so it uses the expanded template files list
+            self._resource_handler = None
+            self._apply_all_resources()
+        # Write to its own application bucket
+        relation_data = self.model.get_relation("otel-collector", event.relation.id).data[self.app]
+        # Update own application bucket with otel collector information
+        relation_data.update(
+            {
+                "otel_collector_svc_namespace": self.model.name,
+                "otel_collector_svc_name": "otel-collector",
+                "otel_collector_port": "55678",
+            }
```

Comment:
There is a race condition here that I don't think a user of the relation would expect.  Because the otel resources take time to deploy, `relation_data.update()` here will, at least for the first relation, send the relation data to all `otel-collector` consumers before the `otel` deployment is up.  So effectively the contract to the consumer is "here is a scraper that **may** be up".  Is knative serving/eventing ok with that?  

This would feel cleaner if data was only written to this relation once the scraper actually existed.  Not that this is *wrong*, but it doesn't feel like what would be expected.

Reply:
If you meant to say "collector" instead of "scraper", then what would happen is KnativeEventing and KnativeServing will change their configuration using whatever relation data they receive, which means they will attempt to communicate to some collector for sending metrics, if they do not find one, the metrics collection will not be successful, but no errors should block eventing/serving components.

Point taken, though, we assume that the otel deployment is up and running when we update the relation data, which could not be the case. The solution to me is to have a check for ensuring not only this but all k8s resources are up and running after we call self._apply_and_set_status(), but then (as we discussed) this seems like a solution that belongs in `chisme`.

Reply:
I have filed an [issue](https://github.com/canonical/knative-operators/issues/60) for making sure we are aware of this behaviour, in the meantime, I am leaving this as is and track possible changes in the aforementioned issue.

Reply:
Since the end consumer to all this (knative eventing/serving) is fault tolerant to this thing not actually being where we say, then this sounds good.

Is there somewhere we could put this that a consumer of this relation could see, that way they know not to rely on it being available?  I can't think of a good place, but if you can that would be nice.  Also, can we link the filed issue in the `_on_otel_collector_relation_created` docstring so someone reading the code knows the issue?  with that extra comment, this lgtm and you can resolve

Reply:
I have added a comment and linked the issue. Not sure what you mean by "Is there somewhere we could put this that a consumer of this relation could see, that way they know not to rely on it being available? "

Reply:
Comment is good enough.  Ideally, there'd be some doc somewhere (like maybe in the metadata.yaml file?) that explains how to use the relation, but I dont think there is a natural spot for this so we can ignore